src/Microsoft.Identity.Client/TokenCacheExtensions.cs

```diff
@@ -100,8 +106,70 @@ public static void Deserialize(this TokenCache tokenCache, byte[] unifiedState)
             GuardOnMobilePlatforms();
             lock (tokenCache.LockObject)
             {
-                RequestContext requestContext = new RequestContext(null, new MsalLogger(Guid.Empty, null));
-                TokenCacheSerializeHelper.DeserializeUnifiedCache(tokenCache.TokenCacheAccessor, unifiedState, requestContext);
+                var requestContext = tokenCache.CreateRequestContext();
+                tokenCache.TokenCacheAccessor.Clear();
+
+                Dictionary<string, IEnumerable<string>> cacheDict = JsonHelper
+                    .DeserializeFromJson<Dictionary<string, IEnumerable<string>>>(unifiedState);
+
+                if (cacheDict == null || cacheDict.Count == 0)
+                {
+                    tokenCache.Logger.Info("Msal Cache is empty.");
+                    return;
+                }
+
+                if (cacheDict.ContainsKey(AccessTokenKey))
+                {
+                    foreach (var atItem in cacheDict[AccessTokenKey])
+                    {
+                        var msalAccessTokenCacheItem =
+                            JsonHelper.TryToDeserializeFromJson<MsalAccessTokenCacheItem>(atItem, requestContext);
+                        if (msalAccessTokenCacheItem != null)
+                        {
+                            tokenCache.TokenCacheAccessor.SaveAccessToken(msalAccessTokenCacheItem);
```

Comment:
>SaveAccessToken [](start = 58, length = 15)

While not really your code... (but since you refactor things)

When it says save... I get a bit nervous. does this cause the external cache events to be fired? 

Perhaps this should rather be called AppendAccessToken ?